src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -581,13 +579,54 @@ private void checkDataLength(int ... lengths) {
         }
     }
 
+    /**
+     * Intrinsic for Vector AES Galois Counter Mode implementation.
+     * AES and GHASH operations are interleaved in the intrinsic implementation.
+     * return - number of processed bytes
+     *
+     * Requires 768 bytes (48 AES blocks) to efficiently use the intrinsic
+     * @param in input buffer
+     * @param inOfs input offset
+     * @param ct buffer that ghash will read (in for encrypt, out for decrypt)
+     * @param ctOfs offset for ct buffer
+     * @param out output buffer
+     * @param outOfs output offset
+     * @param gctr object for the CTR operation
+     * @param ghash object for the ghash operation
+     */
+    @IntrinsicCandidate
+    private static int implGCMCrypt(byte[] in, int inOfs, int inLen,
+        byte[] ct, int ctOfs, byte[] out, int outOfs,
+        GCTR gctr, GHASH ghash) {
+
+        int len = 0;
+        if (inLen > TRIGGERLEN) {
+            int i = 0;
+            int segments = (inLen / 6);
+            segments -= segments % gctr.blockSize;
+            do {
+                len += gctr.update(in, inOfs + len, segments, out,
+                        outOfs + len);
+                ghash.update(ct, ctOfs, segments);
+                ctOfs = len;
+            } while (++i < 5);
+
+            inLen -= len;
+        }
+        len += gctr.update(in, inOfs + len, inLen, out, outOfs + len);
+        ghash.update(ct, ctOfs, inLen);
+        return len;
+    }
+
+
     /**
      * Abstract class for GCMEncrypt and GCMDecrypt internal context objects
      */
     abstract class GCMEngine {
         byte[] preCounterBlock;
-        GCTR gctrPAndC;
-        GHASH ghashAllToS;
+        GCTR gctr;
+        GHASH ghash;
+        GCMOperation op;
```

Comment:
Now that you have GCMOperation op, but there is still if-else blocks checking whether it's encryption/decryption and uses gctr and ghash instead of op. Looks like a bit adhoc? Can GaloisCounterMode.implGCMCrypt(...) just take GCMOperation op instead, no need for ct, ctOfs, gctr and ghash?

Reply:
Initializing op in abstract GCMEngine would mean another 'if(encryption)', when that would not be needed in the  GCMEncrypt() or GCMDecrypt().  I don't see why that is clearer. 

GaloisCounterMode.implGCMCrypt(...) is the intrinsic, so I have to use what is used by hotspot.

Reply:
Seems strange to have GCMOperation op defined in GCMEngine but not initialized, nor used. The methods in GCMEngine which use op has an argument named op anyway. Either you just use the "op" field (remove the "op" argument) or the "op" argument (move the op field to GCMEncrypt/GCMDecrypt class). Having both looks confusing.

Reply:
Ok.. Moving it into GCMEncrypt makes sense.  Now that I look at the code GCMDecrypt only uses it when passed to a method.  GCMEncrypt uses it

Reply:
This is still present in the latest update. Is there another update coming?

Reply:
Yes.